### PR TITLE
Rings as config map

### PR DESCRIPTION
**Pull request by kaweue**

### CODE REVIEW

**Comment by kaweue** [comment]
recheck ha

File: ringcontroller/ring_controller.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    def reconcile(self, devices):
+        def format_device(d):
+            return 'r%(region)sz%(zone)s-%(ip)s:%(port)s/%(device)s' % d
+
+        changed = False
+        # Add new devices
+        for dev_string in devices:
+            dev = parse_add_value(dev_string)
+            if len(self.builder.search_devs(dev)) == 0:
```
Comment (Dysproz) [inline ringcontroller/ring_controller.py:25]: `if not len(..):`

File: ringcontroller/ring_controller_test.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
+import unittest
+import base64
+import io
+from swift.common.ring import RingBuilder
+from swift.common.ring.utils import parse_add_value
+from gzip import decompress
+
+from ring_controller import RingController
```
Comment (Dysproz) [inline ringcontroller/ring_controller_test.py:8]: As before, import alphabetically

File: ringcontroller/ring_controller_test.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+class TestRingController(unittest.TestCase):
+
+    def test_reconcile_add_devices(self):
+        # given a empty ring builder
+        b = RingBuilder(10, 1, 1)
+        r = RingController(b, "object")
+
+        # when new devices are added
+        ds = ["r1z1-192.168.0.2:6000/d3", "r1z2-192.168.2.2:5000/d1"]
+        r.reconcile(ds)
```
Comment (Dysproz) [inline ringcontroller/ring_controller_test.py:20]: If you setup the same thing in every test then check setUp method of unit test class. It executes code that then is common for whole class. Just you'll need to add `self.` before objects that you want to use

File: ringcontroller/ring_controller_test.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+        # then correct serialized builder is returned
+        builder_ = base64.b64decode(ring_data["object"])
+        loadedBuilder = RingBuilder.load(
+            "", open=lambda a, b: io.BytesIO(builder_))
+        self.assertEqual(b.devs, loadedBuilder.devs)
+
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (Dysproz) [inline ringcontroller/ring_controller_test.py:109]: Not sure whether you need that. Generally if filename of test file starts with `test_` then you may run pytest command and it'll automatically detects tests

Comment (Dysproz) [reply ringcontroller/ring_controller_test.py:109]: Oh yeah, because all your tests files end with `_test`. In python convention is to start with `test_`

Comment (kaweue) [reply ringcontroller/ring_controller_test.py:109]: I changed that to test_ring_controller.py but bazel is running those tests only if this expression is there
```if __name__ == '__main__':
    unittest.main()```

File: ringcontroller/main.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    logger.info("loading config_map %s/%s", l[0], l[1])
+    try:
+        config_map = v1.read_namespaced_config_map(
+            namespace=l[0], name=l[1])
+
+    except ApiException as e:
+        print("api error: %s" % e)
+        exit(1)
+
+    return config_map
```
[the remark this answers is not in the document]
Comment (kaweue) [reply ringcontroller/main.py]: improved

File: ringcontroller/main.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def load_config():
+    try:
+        config.load_incluster_config()
+    except ConfigException:
+        config.load_kube_config()
+        pass
```
[the remark this answers is not in the document]
Comment (kaweue) [reply ringcontroller/main.py]: removed pass

File: ringcontroller/main.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        print("each device has format rREGIONzZONE-IP:PORT/DEVICE")
+        exit(1)
+
+    logger.debug(argv)
+
+    load_config()
+    v1 = client.CoreV1Api()
+    config_map_namespace_name = argv[1]
+    ring_type = argv[2]
+    device_strings = argv[3:]
```
[the remark this answers is not in the document]
Comment (kaweue) [reply ringcontroller/main.py]: Used argparse for args parsing

File: ringcontroller/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+python-dateutil==2.8.1
+PyYAML==5.3.1
+requests==2.24.0
+requests-oauthlib==1.3.0
+rsa==4.6
+six==1.15.0
+swift==2.25.0
+urllib3==1.25.9
+websocket-client==0.57.0
+setuptools==39.2.0
```
[the remark this answers is not in the document]
Comment (kaweue) [reply ringcontroller/requirements.txt]: sorted :)

File: ringcontroller/ring_controller.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+from swift.common.ring.utils import parse_add_value
+from gzip import GzipFile
+
+
+class RingController(object):
+    def __init__(self, builder, ring_type, logger=None):
+        self.builder = builder
+        self.ring_type = ring_type
+        self.logger = logger or logging.getLogger('RingController')
```
[the remark this answers is not in the document]
Comment (kaweue) [reply ringcontroller/ring_controller.py]: added _

File: ringcontroller/ring_controller.py (hunk 1 above)
Comment (kaweue) [reply ringcontroller/ring_controller.py]: changed

File: ringcontroller/ring_controller_test.py (hunk 2 above)
Comment (kaweue) [reply ringcontroller/ring_controller_test.py]: sorted

File: ringcontroller/ring_controller_test.py (hunk 3 above)
Comment (kaweue) [reply ringcontroller/ring_controller_test.py]: Actually I prefer not to have setUp method. I think it looks better without it.

File: ringcontroller/ring_controller_test.py (hunk 4 above)
Comment (kaweue) [reply ringcontroller/ring_controller_test.py]: changed name of file but I left this `unittest.main()` without this bazel wasn't executing tests.

File: ringcontroller/ring_controller_test.py (hunk 3 above)
Comment (Dysproz) [reply ringcontroller/ring_controller_test.py]: ok

File: ringcontroller/ring_controller_test.py (hunk 4 above)
Comment (Dysproz) [reply ringcontroller/ring_controller_test.py]: ok

File: ringcontroller/main.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+    logger.debug(args)
+    load_config()
+    v1 = client.CoreV1Api()
+    config_map = load_config_map(v1, args.config_map_name)
+
+    if config_map.binary_data != None and args.ring_type in config_map.binary_data:
+        builder_ = base64.b64decode(config_map.binary_data[args.ring_type])
+        logger.info("loading existing ring builder")
+        builder = RingBuilder.load("", open=lambda a, b: io.BytesIO(builder_))
```
Comment (shifali19) [inline ringcontroller/main.py:76]: You can delete "pkg/swift/ring/reconcile_ring.go" as it will not be used anymore.

Comment (kaweue) [reply ringcontroller/main.py:76]: Ok I will do that in my next PR

**Pull request merged**